File: js/packages/browser/src/transport.js
/**
 * Injectable web transport. MapTransport for offline tests; HttpTransport for live fetch.
 */

export class TransportResponse {
  constructor({
    status = 0,
    finalUrl = "",
    contentType = "",
    body = "",
    headers = {},
    error = "",
    final_url,
    content_type,
  } = {}) {
    this.status = status;
    this.finalUrl = final_url ?? finalUrl;
    this.contentType = content_type ?? contentType;
    this.body = body;
    this.headers = { ...headers };
    this.error = error;
  }

  ok() {
    return !this.error && this.status >= 200 && this.status < 400;
  }

  toDict() {
    return {
      status: this.status,
      final_url: this.finalUrl,
      content_type: this.contentType,
      body: this.body,
      headers: { ...this.headers },
      error: this.error,
    };
  }
}

export class TransportRequest {
  constructor({
    url = "",
    method = "GET",
    headers = {},
    timeoutMs = 15000,
    maxBodyBytes = 2 * 1024 * 1024,
    followRedirects = true,
    maxRedirects = 5,
    timeout_ms,
    max_body_bytes,
    follow_redirects,
    max_redirects,
  } = {}) {
    this.url = url;
    this.method = method;
    this.headers = { ...headers };
    this.timeoutMs = timeout_ms ?? timeoutMs;
    this.maxBodyBytes = max_body_bytes ?? maxBodyBytes;
    this.followRedirects = follow_redirects ?? followRedirects;
    this.maxRedirects = max_redirects ?? maxRedirects;
  }
}

/** Offline / fixture transport. */
export class MapTransport {
  constructor(pages = {}) {
    this._pages = new Map();
    this._errors = new Map();
    for (const [url, value] of Object.entries(pages)) {
      if (typeof value === "string") this.setPage(url, value);
      else this.setPage(url, value.body ?? "", value.status ?? 200, value.contentType);
    }
  }

  name() {
    return "map";
  }

  setPage(url, body, status = 200, contentType = "text/html; charset=utf-8") {
    this._pages.set(url, { body: String(body ?? ""), status, contentType });
    this._errors.delete(url);
    return this;
  }

  setError(url, error) {
    this._errors.set(url, String(error ?? "transport error"));
    this._pages.delete(url);
    return this;
  }

  clear() {
    this._pages.clear();
    this._errors.clear();
    return this;
  }

  async get(request) {
    const req = request instanceof TransportRequest ? request : new TransportRequest(request);
    const url = req.url;
    if (this._errors.has(url)) {
      return new TransportResponse({ error: this._errors.get(url), finalUrl: url });
    }
    const page = this._pages.get(url);
    if (!page) {
      return new TransportResponse({
        status: 404,
        finalUrl: url,
        error: `map transport: no page for ${url}`,
      });
    }
    let body = page.body;
    if (req.maxBodyBytes > 0 && body.length > req.maxBodyBytes) {
      body = body.slice(0, req.maxBodyBytes);
    }
    return new TransportResponse({
      status: page.status,
      finalUrl: url,
      contentType: page.contentType,
      body,
      headers: { "content-type": page.contentType },
    });
  }
}

/** Live HTTP via native fetch (or inject fetchImpl). */
export class HttpTransport {
  constructor({ fetchImpl = null, name: transportName = "http", retries = 2, baseDelayMs = 300 } = {}) {
    this._fetch = fetchImpl;
    this._name = transportName;
    this._retries = Math.max(0, Number(retries) || 0);
    this._baseDelayMs = Math.max(0, Number(baseDelayMs) || 0);
  }

  name() {
    return this._name;
  }

  async get(request) {
    const req = request instanceof TransportRequest ? request : new TransportRequest(request);
    const fetchFn = this._fetch ?? globalThis.fetch;
    if (typeof fetchFn !== "function") {
      return new TransportResponse({
        error: "HttpTransport requires global fetch or fetchImpl",
        finalUrl: req.url,
      });
    }

    let last = null;
    for (let attempt = 0; attempt <= this._retries; attempt++) {
      last = await this._once(fetchFn, req);
      const retryable =
        last.error?.includes("timeout") ||
        last.status === 429 ||
        last.status === 503 ||
        last.status === 502 ||
        last.status === 504;
      if (!retryable || attempt >= this._retries) return last;
      const delay = this._baseDelayMs * 2 ** attempt;
      if (delay > 0) await new Promise((r) => setTimeout(r, delay));
    }
    return last;
  }

  async _once(fetchFn, req) {
    const controller = typeof AbortController !== "undefined" ? new AbortController() : null;
    const timer =
      controller && req.timeoutMs > 0
        ? setTimeout(() => controller.abort(), req.timeoutMs)
        : null;

    try {
      const headers = {
        Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        ...req.headers,
      };
      const response = await fetchFn(req.url, {
        method: req.method || "GET",
        headers,
        redirect: req.followRedirects ? "follow" : "manual",
        signal: controller?.signal,
      });
      const contentType = response.headers?.get?.("content-type") ?? "";
      let body = await response.text();
      if (req.maxBodyBytes > 0 && body.length > req.maxBodyBytes) {
        body = body.slice(0, req.maxBodyBytes);
      }
      const outHeaders = {};
      if (response.headers && typeof response.headers.forEach === "function") {
        response.headers.forEach((value, key) => {
          outHeaders[key] = value;
        });
      }
      return new TransportResponse({
        status: response.status,
        finalUrl: response.url || req.url,
        contentType,
        body,
        headers: outHeaders,
      });
    } catch (error) {
      const message =
        error?.name === "AbortError"
          ? `timeout after ${req.timeoutMs}ms`
          : error instanceof Error
            ? error.message
            : String(error);
      return new TransportResponse({ error: message, finalUrl: req.url });
    } finally {
      if (timer) clearTimeout(timer);
    }
  }
}

export function makeTransport(kind = "http", opts = {}) {
  const key = String(kind ?? "http").toLowerCase();
  if (key === "map" || key === "stub" || key === "offline" || key === "fixture") {
    if (key === "fixture") return makeFixtureMapTransport();
    return new MapTransport();
  }
  if (key === "http" || key === "live" || key === "https") {
    return new HttpTransport(opts);
  }
  throw new TypeError(`unknown transport kind: ${kind}`);
}

export function defaultTransport(preferLive = true) {
  return preferLive ? new HttpTransport() : new MapTransport();
}

export function makeFixtureMapTransport() {
  const map = new MapTransport();
  const index = `<!DOCTYPE html>
<html><head><title>Fixture Home</title></head>
<body>
<h1>Welcome to Fixture</h1>
<p>Home page for offline web explorer tests. Alpha &amp; beta notes.</p>
<script>secret_should_not_appear();</script>
<a href="/about.html">About Us</a>
<a href="/docs/guide.html">Guide</a>
<a href="https://evil.example/block-me">External Evil</a>
</body></html>`;
  const about = `<html><head><title>About Fixture</title></head>
<body><p>About page content with more detail.</p>
<a href="/">Home</a>
<a href="/docs/guide.html">Guide</a>
</body></html>`;
  const guide = `<html><head><title>Guide</title></head>
<body><h2>User Guide</h2><p>Step one: configure ExplorePolicy. Step two: inject WebTransport.</p>
<a href="/">Home</a>
</body></html>`;
  map.setPage("https://fixture.local/", index);
  map.setPage("https://fixture.local/index.html", index);
  map.setPage("https://fixture.local/about.html", about);
  map.setPage("https://fixture.local/docs/guide.html", guide);
  map.setPage("https://fixture.local/missing.html", "", 404);
  map.setError("https://fixture.local/boom", "simulated transport failure");
  return map;
}

File: packages/python/handoffkit/browser/transport.py
"""Injectable web transports (map / http / fixture)."""

from __future__ import annotations

import time
import urllib.error
import urllib.request
from dataclasses import dataclass, field
from datetime import datetime, timezone
from email.utils import parsedate_to_datetime
from typing import Any, Protocol

from handoffkit.browser.types import DEFAULT_UA


def _retry_after_ms(headers: dict[str, str]) -> int:
    raw = str(headers.get("retry-after") or headers.get("Retry-After") or "").strip()
    if not raw:
        return 0
    if raw.isdigit():
        return max(0, int(raw) * 1000)
    try:
        when = parsedate_to_datetime(raw)
        if when.tzinfo is None:
            when = when.replace(tzinfo=timezone.utc)
        delta = (when - datetime.now(timezone.utc)).total_seconds()
        return max(0, int(delta * 1000))
    except (TypeError, ValueError, OverflowError):
        return 0


@dataclass
class TransportResponse:
    status: int = 0
    final_url: str = ""
    content_type: str = ""
    body: str = ""
    headers: dict[str, str] = field(default_factory=dict)
    error: str = ""
    retry_after_ms: int = 0

    def ok(self) -> bool:
        return not self.error and 200 <= self.status < 400

    def to_dict(self) -> dict[str, Any]:
        return {
            "status": self.status,
            "final_url": self.final_url,
            "content_type": self.content_type,
            "body": self.body,
            "headers": dict(self.headers),
            "error": self.error,
            "retry_after_ms": self.retry_after_ms,
        }


class WebTransport(Protocol):
    def name(self) -> str: ...

    def get(
        self,
        url: str,
        timeout_ms: int = 15000,
        headers: dict[str, str] | None = None,
        max_body_bytes: int = 2 * 1024 * 1024,
    ) -> TransportResponse: ...


class MapTransport:
    def __init__(self, pages: dict[str, Any] | None = None) -> None:
        self._pages: dict[str, tuple[str, int, str]] = {}
        self._errors: dict[str, str] = {}
        for url, value in (pages or {}).items():
            if isinstance(value, str):
                self.set_page(url, value)
            else:
                self.set_page(
                    url,
                    value.get("body", ""),
                    int(value.get("status", 200)),
                    value.get("content_type", "text/html; charset=utf-8"),
                )

    def name(self) -> str:
        return "map"

    def set_page(
        self,
        url: str,
        body: str,
        status: int = 200,
        content_type: str = "text/html; charset=utf-8",
    ) -> MapTransport:
        self._pages[url] = (str(body or ""), status, content_type)
        self._errors.pop(url, None)
        return self

    def set_error(self, url: str, error: str) -> MapTransport:
        self._errors[url] = str(error or "transport error")
        self._pages.pop(url, None)
        return self

    def clear(self) -> MapTransport:
        self._pages.clear()
        self._errors.clear()
        return self

    def get(
        self,
        url: str,
        timeout_ms: int = 15000,
        headers: dict[str, str] | None = None,
        max_body_bytes: int = 2 * 1024 * 1024,
    ) -> TransportResponse:
        _ = timeout_ms, headers
        if url in self._errors:
            return TransportResponse(final_url=url, error=self._errors[url])
        page = self._pages.get(url)
        if page is None:
            return TransportResponse(
                status=404,
                final_url=url,
                error=f"map transport: no page for {url}",
            )
        body, status, ctype = page
        if max_body_bytes > 0 and len(body) > max_body_bytes:
            body = body[:max_body_bytes]
        return TransportResponse(
            status=status,
            final_url=url,
            content_type=ctype,
            body=body,
            headers={"content-type": ctype},
        )


class HttpTransport:
    def __init__(self, *, retries: int = 2, base_delay_ms: int = 300) -> None:
        self._retries = max(0, int(retries))
        self._base_delay_ms = max(0, int(base_delay_ms))

    def name(self) -> str:
        return "http"

    def get(
        self,
        url: str,
        timeout_ms: int = 15000,
        headers: dict[str, str] | None = None,
        max_body_bytes: int = 2 * 1024 * 1024,
    ) -> TransportResponse:
        last = TransportResponse(final_url=url, error="no attempt")
        for attempt in range(self._retries + 1):
            last = self._once(url, timeout_ms, headers, max_body_bytes)
            retryable = "timeout" in (last.error or "").lower() or last.status in {
                429,
                502,
                503,
                504,
            }
            if not retryable or attempt >= self._retries:
                return last
            delay = self._base_delay_ms * (2**attempt) / 1000.0
            if last.retry_after_ms > 0:
                delay = max(delay, last.retry_after_ms / 1000.0)
            if delay > 0:
                time.sleep(min(delay, 30.0))
        return last

    def _once(
        self,
        url: str,
        timeout_ms: int,
        headers: dict[str, str] | None,
        max_body_bytes: int,
    ) -> TransportResponse:
        req_headers = {
            "User-Agent": DEFAULT_UA,
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        }
        if headers:
            req_headers.update(headers)
        request = urllib.request.Request(url, headers=req_headers, method="GET")
        try:
            with urllib.request.urlopen(request, timeout=max(timeout_ms / 1000.0, 1.0)) as resp:
                raw = resp.read()
                if max_body_bytes > 0 and len(raw) > max_body_bytes:
                    raw = raw[:max_body_bytes]
                charset = resp.headers.get_content_charset() or "utf-8"
                body = raw.decode(charset, errors="replace")
                out_headers = {k.lower(): v for k, v in resp.headers.items()}
                return TransportResponse(
                    status=getattr(resp, "status", 200) or 200,
                    final_url=resp.geturl(),
                    content_type=resp.headers.get("Content-Type", ""),
                    body=body,
                    headers=out_headers,
                    retry_after_ms=_retry_after_ms(out_headers),
                )
        except urllib.error.HTTPError as exc:
            raw = exc.read() if hasattr(exc, "read") else b""
            if max_body_bytes > 0 and len(raw) > max_body_bytes:
                raw = raw[:max_body_bytes]
            body = raw.decode("utf-8", errors="replace") if raw else ""
            err_headers = {k.lower(): v for k, v in (exc.headers.items() if exc.headers else [])}
            return TransportResponse(
                status=int(exc.code),
                final_url=url,
                body=body,
                headers=err_headers,
                error=f"HTTP status {exc.code}",
                retry_after_ms=_retry_after_ms(err_headers),
            )
        except Exception as exc:  # noqa: BLE001
            return TransportResponse(final_url=url, error=str(exc))


def make_transport(kind: str = "http") -> MapTransport | HttpTransport:
    key = str(kind or "http").lower()
    if key in {"map", "stub", "offline"}:
        return MapTransport()
    if key == "fixture":
        return make_fixture_map_transport()
    if key in {"http", "live", "https"}:
        return HttpTransport()
    raise TypeError(f"unknown transport kind: {kind}")


def default_transport(prefer_live: bool = True) -> MapTransport | HttpTransport:
    return HttpTransport() if prefer_live else MapTransport()


def make_fixture_map_transport() -> MapTransport:
    m = MapTransport()
    index = (
        "<!DOCTYPE html><html><head><title>Fixture Home</title></head><body>"
        "<h1>Welcome to Fixture</h1>"
        "<p>Home page for offline web explorer tests. Alpha &amp; beta notes.</p>"
        "<script>secret_should_not_appear();</script>"
        '<a href="/about.html">About Us</a>'
        '<a href="/docs/guide.html">Guide</a>'
        '<a href="https://evil.example/block-me">External Evil</a>'
        "</body></html>"
    )
    about = (
        "<html><head><title>About Fixture</title></head><body>"
        "<p>About page content with more detail.</p>"
        '<a href="/">Home</a><a href="/docs/guide.html">Guide</a>'
        "</body></html>"
    )
    guide = (
        "<html><head><title>Guide</title></head><body>"
        "<h2>User Guide</h2>"
        "<p>Step one: configure ExplorePolicy. Step two: inject WebTransport.</p>"
        '<a href="/">Home</a></body></html>'
    )
    m.set_page("https://fixture.local/", index)
    m.set_page("https://fixture.local/index.html", index)
    m.set_page("https://fixture.local/about.html", about)
    m.set_page("https://fixture.local/docs/guide.html", guide)
    m.set_page("https://fixture.local/missing.html", "", 404)
    m.set_error("https://fixture.local/boom", "simulated transport failure")
    return m
